java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightStream.java

```diff
@@ -131,33 +130,53 @@ public DictionaryProvider takeDictionaryOwnership() {
     return provider;
   }
 
+  /**
+   * Get the descriptor for this stream. Only applicable on the server side of a DoPut operation. Will block until the
+   * client sends the descriptor.
+   */
   public FlightDescriptor getDescriptor() {
+    // This blocks until the schema message (with the descriptor) is sent.
+    getRoot();
     return descriptor;
   }
 
   /**
    * Closes the stream (freeing any existing resources).
    *
-   * <p>If the stream is isn't complete and is cancellable this method will cancel the stream first.</p>
+   * <p>If the stream isn't complete and is cancellable, this method will cancel the stream first.</p>
    */
   public void close() throws Exception {
-    if (!completed && cancellable != null) {
-      cancel("Stream closed before end.", null);
+    RuntimeException cancelException = null;
+    try {
+      // cancel() can throw, but we still want to clean up resources.
+      if (!completed && cancellable != null) {
+        cancel("Stream closed before end.", /* no exception to report */ null);
+      }
+    } catch (RuntimeException e) {
+      cancelException = e;
+    }
+
+    try {
+      final ImmutableList.Builder<AutoCloseable> closeables = ImmutableList.builder();
```

Comment:
This feels a bit wonky. Seems like there would be a better way to do it. Can we do this by concating a set of lists. For example, why not use queue.stream() instead of Arrays.stream(queue.toArray()). Also, why do we need the getClass()? How about instead we just add only AutoCloseable items to the list (just make the sentinel closeable).

Also, AutoCloseables.close() shouldn't mind having null values (I think) so you should be able to skip the null check.

These comments are all cosmetic.

Reply:
The null check is in fact because `ImmutableList` doesn't accept nulls. You're right, `AutoCloseables.close` is fine with null.

I reworked this to use a plain `ArrayList`, and changed to use AutoCloseable sentinel objects. That way we can simplify this method drastically.